src/core/ipns/routing/pubsub-datastore.js

```diff
@@ -0,0 +1,171 @@
+    this._pubsubDs = new PubsubDatastore(pubsub, localDatastore, peerId, ipns.validator, this._handleSubscriptionKey)
+  }
+
+  /**
+   * Put a value to the pubsub datastore indexed by the received key properly encoded.
+   * @param {Buffer} key identifier of the value.
+   * @param {Buffer} value value to be stored.
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  put (key, value, callback) {
+    if (!Buffer.isBuffer(key)) {
+      const errMsg = `key ${key} does not have a valid format`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    if (!Buffer.isBuffer(value)) {
+      const errMsg = `received value ${value} is not a buffer`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_VALUE_RECEIVED'))
+    }
+
+    this._pubsubDs.put(key, value, callback)
+  }
+
+  /**
+   * Get a value from the pubsub datastore indexed by the received key properly encoded.
+   * Moreover, the identifier topic is subscribed and the pubsub datastore records will be
+   * updated once new publishes occur.
+   * @param {Buffer} key identifier of the value to be obtained.
+   * @param {function(Error, Buffer)} callback
+   * @returns {void}
+   */
+  get (key, callback) {
+    if (!Buffer.isBuffer(key)) {
+      const errMsg = `key ${key} does not have a valid format`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    this._pubsubDs.get(key, (err, res) => {
+      // Add topic subscribed
+      const ns = key.slice(0, ipnsNSLength)
+
+      if (ns.toString() === ipnsNS) {
+        const stringifiedTopic = key.toString()
+        const id = toB58String(key.slice(ipnsNSLength))
+
+        this._subscriptions[stringifiedTopic] = id
+
+        log(`subscribed pubsub ${stringifiedTopic}: ${id}`)
+      }
+
+      // If no data was obtained, after storing the subscription, return the error.
+      if (err) {
+        return callback(err)
+      }
+
+      callback(null, res)
+    })
+  }
+
+  // Modify subscription key to have a proper encoding
+  _handleSubscriptionKey (key, callback) {
+    const subscriber = this._subscriptions[key]
+
+    if (!subscriber) {
+      const errMsg = `key ${key} does not correspond to a subscription`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    let keys
+    try {
+      keys = ipns.getIdKeys(fromB58String(subscriber))
+    } catch (err) {
+      log.error(err)
+      return callback(err)
+    }
+
+    callback(null, keys.routingKey.toBuffer())
+  }
+
+  /**
+   * Get pubsub subscriptions related to ipns.
+   * @param {function(Error, Object)} callback
+   * @returns {void}
+   */
+  getSubscriptions (callback) {
+    const subscriptions = Object.values(this._subscriptions)
+
+    return callback(null, subscriptions.map((sub) => `/ipns/${sub}`))
```

Comment:
use ipnsNS const here
```suggestion
    return callback(null, subscriptions.map((sub) => `${ipnsNs}${sub}`))
```